src/common/snippets/src/lowered/pass/define_buffer_clusters.cpp

```diff
@@ -42,141 +80,139 @@ size_t DefineBufferClusters::get_cluster_buffer_id(const BufferCluster& cluster)
-                continue;
-
             // Memory can be reused if reading and writing are executed proportionally:
-            //  - the same reading/writing order
-            //  - the same buffer memory sizes
-            if ((input_buffer_expr->get_byte_size() != output_buffer_expr->get_byte_size()) ||
-                (input_buffer_expr->get_output_port_descriptor(0)->get_layout() != output_buffer_expr->get_input_port_descriptor(0)->get_layout()))
+            //  - output buffer can have precision with data size less than input buffer
+            if ((input_buffer_expr->get_data_type().size() < output_buffer_expr->get_data_type().size()))
                 continue;
 
-            // Also memory can be reused if there are the same ShiftPtrParams (data size, final offsets, ptr increments)
-            const auto& input_buffer_ports = in.second;
-            for (const auto& input_buffer_port_idx : input_buffer_ports) {
-                const auto input_params =
-                    ShiftPtrParams(data_sizes[input_buffer_port_idx], ptr_increments[input_buffer_port_idx], final_offsets[input_buffer_port_idx]);
-                const auto output_params =
-                    ShiftPtrParams(data_sizes[output_buffer_port_idx], ptr_increments[output_buffer_port_idx], final_offsets[output_buffer_port_idx]);
-
-                // If data pointer shift parameters are unknown on model compilation stage (dynamic),
-                // we cannot be sure that these data pointers will be proportionally shifted in runtime.
-                if (input_params.is_static() && output_params.is_static() && input_params == output_params) {
-                    const auto cluster_it = find_cluster_by_expr(input_buffer_expr);
-                    OPENVINO_ASSERT(cluster_it != m_clusters.end(), "Buffer on inputs of Loop must be already saved in clusters");
-                    // Add to the existing cluster
-                    has_been_added = cluster_it->insert(output_buffer_expr).second;
-                    OPENVINO_ASSERT(has_been_added, "Buffer has not been saved in cluster");
-                    // Remove input buffer because we have already use its memory
-                    visited_buffers.insert(input_buffer_expr);
-                    break;
-                }
+            //  - Memory can be shared if Buffer has the same allocation size.
+            //    If they're undefined, check that they have the same rule for initialization of allocation size
+            if (input_buffer_expr->is_defined() && output_buffer_expr->is_defined()) {
+                if (input_buffer_expr->get_allocation_size() != output_buffer_expr->get_allocation_size())
+                    continue;
+            } else {
+                if (input_buffer_expr->get_type_info() != BufferExpression::get_type_info_static() ||
+                    output_buffer_expr->get_type_info() != BufferExpression::get_type_info_static())
+                    continue;
```

Comment:
Right! I didn't find about another way to compare them.
We usually derive from `BufferExpression` if the needed buffer should have another allocation size calculation rule.

Reply:
This is OK for now, because it handles all the existing cases. However, the current design allows to derive from `BufferExpression` and not override the `init_allocation_size`: in this case, 2 classes with equal `init_allocation_size` will obviously have 2 different type info, but have the same `init_allocation_size`. And this worries me, since in the described scenario we will get not obvious issue: DefineBufferClusters will work unexpectedly for these 2 buffers only in dynamic scenario.

As an option, we could use C++ RTTI here (if this is even legal ;) ) to detect if 2 classes has the same overloads for the function:
```
template <class T1, class T2>
bool areOverloadsEqual(T1* obj1, T2* obj2) {
    return typeid(&T1::init_allocation_size) == typeid(&T2::init_allocation_size);
}
```
Such check can be used at least in 2 ways:

1. To replace the current type info check
2. As an additional assertion, which contains the following check: "if type_info are different, `init_allocation_size` overloads must be different too"

@IvanNovoselov @a-sidorova what do you think?

Reply:
I fully understand your concern. Moreover, I have the same feelings. I tried to find the solution which resolve our problem - but didn't find..

As for your suggestion, unfortunately it doesn't work because `T` is always `BufferExpression` in our case: all buffers are stored as `std::shared_ptr<BufferExpression>` in the map (or in another collection). So we will try to compare `return typeid(&BufferExpression::init_allocation_size) == typeid(&BufferExpression::init_allocation_size)` - is always `true`.

Need to say that the common algorithm which is implemented here, covers the simple case with Loops and eltwise ops inside. In these cases we always use the base class `BufferExpression`. For now I cannot imagine the case where there are Buffers with another type (but not overrided `init_allocation_size`), connected to one Loop with eltwise ops inside and which should be in one cluster.

So I'd prefer to leave it so since it seems like there is no solution to figure out "overridded the method or not" in runtime.

Reply:
1) We definitely need a comment on why we need this check, and what exactly we are trying to check here
2) The problem here as I see it is that we've lost type info information when we decided to use base class pointers. A standard approach here is to try to downcast to known derived classes, we can use `is_type` here for custom RTTI.
3) Another solution is to employ virtual functions. The thing is that virtual functions work polymorphically even when called through the base pointer. So if we create a virtual function that returns static type info, we'll be able to access it and compare buffer types using only base class pointers. It's easier to explain in code, please check out this demo: [https://coliru.stacked-crooked.com/a/36ac0d5cc4eaca61](https://coliru.stacked-crooked.com/a/36ac0d5cc4eaca61)

So I think we can leave as-is, if we need it only here and never again, but it's worth to create a ticket where a more scalable solution is described.

Reply:
> The problem is that derived class from BufferExpression can have the same BufferExpression::init_allocation_size (not overriden). Then the current check doesn't cover this case.

There is no way to check if a particular virtual function was overriden by the derived class in C++, because the virtual functions table is generated by a compiler and is deliberately hidden from users.
So basically `get_type_info` is our only option (apart from `virtual bool is_init_allocation_size_overriden()` of couse)

Reply:
Thanks for the details! Now we have a bit stricter check than before: both buffers must have not only equal types, but be `BufferExpression` (not inheritor).
`const auto is_the_same_init_algo = utils::everyone_is(BufferExpression::get_type_info_static(), input_buffer_expr->get_type_info(), output_buffer_expr->get_type_info());`

And we do that since this logic is not expected to use for `BufferExpression` inherited classes. I've just written that to make sure that I understood all correctly, so please correct me if I am wrong 😄